kv/index.go

```diff
@@ -0,0 +1,537 @@
+// of a value in a source bucket.
+type IndexSourceOnFunc func([]byte) ([]byte, error)
+
+type indexMapping struct {
+	source []byte
+	index  []byte
+	fn     IndexSourceOnFunc
+}
+
+func (i indexMapping) SourceBucket() []byte { return i.source }
+
+func (i indexMapping) IndexBucket() []byte { return i.index }
+
+func (i indexMapping) IndexSourceOn(v []byte) ([]byte, error) {
+	return i.fn(v)
+}
+
+// NewIndexMapping creates an implementation of IndexMapping for the provided source bucket
+// to a destination index bucket.
+func NewIndexMapping(sourceBucket, indexBucket []byte, fn IndexSourceOnFunc) IndexMapping {
+	return indexMapping{
+		source: sourceBucket,
+		index:  indexBucket,
+		fn:     fn,
+	}
+}
+
+// NewIndex configures and returns a new *Index for a given index mapping.
+// By default the read path (Walk) is disabled. This is because the index needs to
+// be fully populated before depending upon the read path.
+// The read path can be enabled using WithIndexReadPathEnabled option.
+func NewIndex(mapping IndexMapping, opts ...IndexOption) *Index {
+	index := &Index{
+		IndexMapping:      mapping,
+		populateBatchSize: defaultPopulateBatchSize,
+	}
+
+	for _, opt := range opts {
+		opt(index)
+	}
+
+	return index
+}
+
+// IndexPopulatorStore is a store which also has a AutoPopulateIndex method
+// The method returns true when the store support auto population of index
+// on initialize.
+type IndexPopulatorStore interface {
+	Store
+	AutoPopulateIndex() bool
+}
+
+// Initialize creates the index bucket on the provided store
+func (i *Index) Initialize(ctx context.Context, store Store) error {
+	if err := store.Update(ctx, func(tx Tx) error {
+		_, err := i.indexBucket(tx)
+		return err
+	}); err != nil {
+		return err
+	}
+
+	if store, ok := store.(IndexPopulatorStore); ok && store.AutoPopulateIndex() {
+		_, err := i.Populate(ctx, store)
+		return err
+	}
+
+	return nil
+}
+
+func (i *Index) indexBucket(tx Tx) (Bucket, error) {
+	return tx.Bucket(i.IndexBucket())
+}
+
+func (i *Index) sourceBucket(tx Tx) (Bucket, error) {
+	return tx.Bucket(i.SourceBucket())
+}
+
+func indexKey(foreignKey, primaryKey []byte) (newKey []byte, _ error) {
+	newKey = make([]byte, len(primaryKey)+len(foreignKey)+1)
+	copy(newKey, foreignKey)
+	newKey[len(foreignKey)] = '/'
+	copy(newKey[len(foreignKey)+1:], primaryKey)
+
+	return
+}
+
+func indexKeyParts(indexKey []byte) (fk, pk []byte, err error) {
+	// this function is called with items missing in index
+	parts := bytes.SplitN(indexKey, []byte("/"), 2)
+	if len(parts) < 2 {
+		return nil, nil, errors.New("malformed index key")
+	}
+
+	// parts are fk/pk
+	fk, pk = parts[0], parts[1]
+
+	return
+}
+
+// Insert creates a single index entry for the provided primary key on the foreign key.
+func (i *Index) Insert(tx Tx, foreignKey, primaryKey []byte) error {
+	newKey, err := indexKey(foreignKey, primaryKey)
+	if err != nil {
+		return err
+	}
```

Comment:
nit: As mentioned earlier, by changing the `indexKey` return values, you can remove the redundant `if err != nil` check, given `indexKey` always returns `nil` for the `error` value.